packages/@ember/-internals/metal/lib/chain-tags.ts

```diff
@@ -70,48 +70,71 @@ export function getChainTagsForKey(obj: any, path: string) {
 
     segment = path.slice(lastSegmentEnd, segmentEnd);
 
-    // If the segment is an @each, we can process it and then opt-
+    // If the segment is an @each, we can process it and then break
     if (segment === '@each' && segmentEnd !== pathLength) {
-      assert(
-        `When using @each, the value you are attempting to watch must be an array, was: ${current.toString()}`,
-        Array.isArray(current) || isEmberArray(current)
-      );
-
       lastSegmentEnd = segmentEnd + 1;
       segmentEnd = path.indexOf('.', lastSegmentEnd);
 
-      // There shouldn't be any more segments after an `@each`, so break
+      // There should be exactly one segment after an `@each` (i.e. `@each.foo`, not `@each.foo.bar`)
       deprecate(
-        `When using @each, you can only chain one property level deep, but ${path} contains a nested chain. Please create an intermediary computed property or switch to tracked properties.`,
+        `When using @each in a dependent-key or an observer, ` +
+          `you can only chain one property level deep after ` +
+          `the @each. That is, \`${path.slice(0, segmentEnd)}\` ` +
+          `is allowed but \`${path}\` (which is what you passed) ` +
+          `is not.\n\n` +
+          `This was never supported. Currently, the extra segments ` +
+          `are silently ignored, i.e. \`${path}\` behaves exactly ` +
+          `the same as \`${path.slice(0, segmentEnd)}\`. ` +
+          `In the future, this will throw an error.\n\n` +
+          `If the current behavior is acceptable for your use case, ` +
+          `please remove the extraneous segments by changing your ` +
+          `key to \`${path.slice(0, segmentEnd)}\`. ` +
+          `Otherwise, please create an intermediary computed property ` +
+          `or switch to using tracked properties.`,
         segmentEnd === -1,
         {
           until: '3.17.0',
           id: 'ember-metal.computed-deep-each',
         }
       );
 
-      if (segmentEnd === -1) {
-        segmentEnd = pathLength;
-      }
+      let arrLength = current.length;
 
-      segment = path.slice(lastSegmentEnd, segmentEnd)!;
+      if (
+        typeof arrLength !== 'number' ||
+        // TODO: should the second test be `isEmberArray` instead?
+        !(Array.isArray(current) || 'objectAt' in current)
+      ) {
+        // If the current object isn't an array, there's nothing else to do,
+        // we don't watch individual properties. Break out of the loop.
+        break;
+      } else if (arrLength === 0) {
+        // Fast path for empty arrays
+        chainTags.push(tagForProperty(current, '[]'));
+        break;
+      }
 
-      if (segment.indexOf('.') !== -1) {
```

Comment:
I believe this was dead code.

There are two cases on L93: `segmentEnd === -1` (`foo.@each.bar`) and `segmentEnd !== -1` (`foo.@each.bar.baz`, deprecated).

For the first case, `segment` is `path.slice(lastSegmentEnd, pathLength)` which would be `bar`. It is impossible for this to contain a `.` because we already checked previously.

For the second case, `segment` is `path.slice(lastSegmentEnd, segmentEnd)` which would also be `bar`. It is impossible for this to contain a `.`, because `segmentEnd` is the next occurrence of a `.` after the `@each.`, returned from `indexOf('.', lastSegmentEnd)`.

So in both cases, this conditional cannot ever by true so the branch is effectively dead but we are still doing part of the work (`segment.indexOf('.')`) for no reason.